app/src/main/java/com/nutomic/syncthingandroid/model/Completion.java

```diff
@@ -0,0 +1,175 @@
+ * according to syncthing's REST "/completion" JSON result schema.
+ * Completion model of syncthing's web UI is completion[deviceId][folderId]
+ */
+public class Completion {
+
+    private static final String TAG = "Completion";
+
+    DefaultHashMap<String, DefaultHashMap<String, CompletionInfo>> deviceFolderMap =
+        new DefaultHashMap<String, DefaultHashMap<String, CompletionInfo>>();
+
+    DefaultHashMap<String, CompletionInfo> defaultDevice = new DefaultHashMap<String, CompletionInfo>();
+    CompletionInfo defaultFolder = new CompletionInfo();
+
+    /**
+     * Adds a device to the cache model if it does not exist.
+     */
+    /*
+    private void addDevice(String deviceId) {
+        if (!deviceFolderMap.containsKey(deviceId)) {
+            deviceFolderMap.put(deviceId, defaultDevice);
+        }
+    }
+    */
+
+    /**
+     * Removes a device from the cache model.
+     */
+    /*
+    private void removeDevice(String deviceId) {
+        if (deviceFolderMap.containsKey(deviceId)) {
+            deviceFolderMap.remove(deviceId);
+        }
+    }
+    */
+
+    /**
+     * Adds a folder to the cache model if it does not exist.
+     */
+    private void addFolder(String deviceId, String folderId) {
+        // Add device parent node if it does not exist.
+        if (!deviceFolderMap.containsKey(deviceId)) {
+            deviceFolderMap.put(deviceId, defaultDevice);
+        }
+        // Add folder.
+        if (!deviceFolderMap.get(deviceId).containsKey(folderId)) {
+            deviceFolderMap.get(deviceId).put(folderId, defaultFolder);
+        }
+    }
+
+    /**
+     * Removes a folder from the cache model.
+     */
+    private void removeFolder(String folderId) {
+        for (String deviceId : deviceFolderMap.keySet()) {
+            if (deviceFolderMap.get(deviceId).containsKey(folderId)) {
+                deviceFolderMap.get(deviceId).remove(folderId);
+                break;
+            }
+        }
+    }
+
+    /**
+     * Updates device and folder information in the cache model
+     * after a config update.
+     */
+    public void updateFromConfig(List<Device> newDeviceSet, List<Folder> newFolderSet) {
+        // Handle devices that were removed from the config.
+        List<String> removedDevices = new ArrayList<>();;
+        Boolean deviceFound;
+        for (String deviceId : deviceFolderMap.keySet()) {
+            deviceFound = false;
+            for (Device device : newDeviceSet) {
+                if (device.deviceID.equals(deviceId)) {
+                    deviceFound = true;
+                    break;
+                }
+            }
+            if (!deviceFound) {
+                removedDevices.add(deviceId);
+            }
+        }
+        for (String deviceId : removedDevices) {
+            Log.v(TAG, "updateFromConfig: Remove device '" + deviceId + "' from cache model");
+            deviceFolderMap.remove(deviceId);
+        }

```

Comment:
This block could live inside the `if (!deviceFound)` and remove the removedDevices

Reply:
No, that was the case in an unsubmitted debug version I compiled and I got a java exception complaining about changing the map while it is accessed in a for iterator.